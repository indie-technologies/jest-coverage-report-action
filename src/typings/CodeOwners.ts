export interface CodeOwners {
    team: string;
    paths: string[];
}
