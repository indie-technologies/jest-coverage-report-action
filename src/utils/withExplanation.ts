import { CodeOwners } from './../typings/CodeOwners';
export const withExplanation = (
    text: string,
    explanation: string,
    codeOwner?: CodeOwners
) => {
    const strings: string[] = [];

    if (codeOwner) {
        strings.push(`<p>Coverage for ${codeOwner.team}</p>`);
    }

    strings.push(
        `<div title="${explanation}">${text}<sup>:grey_question:</sup></div>`
    );

    return strings.join('');
};
