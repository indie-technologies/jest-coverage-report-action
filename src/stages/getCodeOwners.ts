import { CodeOwners } from './../typings/CodeOwners';
import { getReportTag } from '../constants/getReportTag';
import { GITHUB_MESSAGE_SIZE_LIMIT } from '../constants/GITHUB_MESSAGE_SIZE_LIMIT';
import { formatCoverage } from '../format/formatCoverage';
import { formatErrors } from '../format/formatErrors';
import { formatRunReport } from '../format/formatRunReport';
import { formatThresholdResults } from '../format/formatThresholdResults';
import { getFailureDetails } from '../format/getFailureDetails';
import { getTestRunSummary } from '../format/summary/getTestRunSummary';
import template from '../format/template.md';
import { JsonReport } from '../typings/JsonReport';
import { Options } from '../typings/Options';
import {SummaryReport, TestRunReport, FailReason} from '../typings/Report';
import { ThresholdResult } from '../typings/ThresholdResult';
import { DataCollector } from '../utils/DataCollector';
import { i18n } from '../utils/i18n';
import { insertArgs } from '../utils/insertArgs';
import {context} from '@actions/github';
import {readFile} from 'fs-extra';
import {ActionError} from '../typings/ActionError';

export const getCodeOwners = async (
    dataCollector: DataCollector<JsonReport>,
    options: Options,
): Promise<CodeOwners[] | undefined> => {
  if (!options.codeOwnersDirectory) {
    return undefined;
  }

  const pathToCodeOwnersFile = options.codeOwnersDirectory;

  try {
    dataCollector.info(
        i18n('loadingCoverageFromFile', {
            pathToCodeOwnersFile,
        })
    );

    const outputBuffer = await readFile(pathToCodeOwnersFile);

    const codeOwnersAsString: string = outputBuffer.toString();
    const codeOwnersByLine = codeOwnersAsString.split("\n");
    const filtered = codeOwnersByLine.filter((line) => line.startsWith('/'));
    const codeOwnersMap: Record<string, string[]> = {};
    filtered.forEach((line) => {
      const [path, team] = line.split(' ');
      codeOwnersMap[team] ||= [];
      codeOwnersMap[team] = codeOwnersMap[team].concat(path);
    });

    return Object.keys(codeOwnersMap).map(key => {
      return {
        team: key,
        paths: codeOwnersMap[key]
      }
    });
} catch (error) {
    if ((error as NodeJS.ErrnoException).code === 'ENOENT') {
        throw new ActionError(FailReason.REPORT_NOT_FOUND, {
            coveragePath: pathToCodeOwnersFile,
        });
    }

    throw new ActionError(FailReason.READING_COVERAGE_FILE_FAILED, {
        error: (error as Error).toString(),
    });
}};
