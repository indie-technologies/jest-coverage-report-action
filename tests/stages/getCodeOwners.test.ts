import { getCodeOwners } from './../../src/stages/getCodeOwners';
import * as all from '@actions/github';

import { createReport, getSha } from '../../src/stages/createReport';
import { JsonReport } from '../../src/typings/JsonReport';
import { Options } from '../../src/typings/Options';
import { createDataCollector } from '../../src/utils/DataCollector';
import report from '../mock-data/jsonReport.json';
import {readFile} from 'fs-extra';

const { mockContext, clearContextMock } = all as any;

const clearMocks = () => {
  (readFile as jest.Mock<any, any>).mockClear();
};

beforeEach(clearMocks);

const DEFAULT_OPTIONS: Options = {
    token: '',
    testScript: '',
    iconType: 'emoji',
    annotations: 'all',
    packageManager: 'npm',
    skipStep: 'all',
    prNumber: 5,
    pullRequest: {
        number: 5,
        head: { sha: '123456', ref: '123' },
        base: { ref: '456' },
    },
    output: ['comment'],
};

describe('getCodeOwners', () => {
    it('should match snapshots', async () => {
      const dataCollector = createDataCollector<JsonReport>();
      dataCollector.add(report);

      (readFile as jest.Mock<any, any>).mockImplementationOnce(() => `
* some-team

# some-comment
/jest/examples/typescript/sum.ts team-one
/jest/examples/typescript/calc.ts team-two
/jest/examples/typescript/sum.js team-one
`);

      expect(
          await getCodeOwners(
              dataCollector,
              DEFAULT_OPTIONS,
          )
      ).toMatchSnapshot();

      expect(
        await getCodeOwners(
            dataCollector,
            {
                ...DEFAULT_OPTIONS,
                codeOwnersDirectory: 'custom directory',
            },
        )
      ).toMatchSnapshot();
    });
});
