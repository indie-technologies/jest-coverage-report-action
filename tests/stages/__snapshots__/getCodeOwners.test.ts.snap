// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getCodeOwners should match snapshots 1`] = `undefined`;

exports[`getCodeOwners should match snapshots 2`] = `
Array [
  Object {
    "paths": Array [
      "/jest/examples/typescript/sum.ts",
      "/jest/examples/typescript/sum.js",
    ],
    "team": "team-one",
  },
  Object {
    "paths": Array [
      "/jest/examples/typescript/calc.ts",
    ],
    "team": "team-two",
  },
]
`;
